### DISCUSSION TITLE
Question about the data migration

### DISCUSSION THREAD

**Discussion by hy2014** [apache/kvrocks#2471] [Q&A] [answered]
i use kvrocks in 2022, and left some data.
for some reason i need move the data from cloud to my local. i build the kvrocks, and copy 2022's data to new kvrocks application. but i found new kvrocks compact my 2022's data and remove all the data. i should i keep my data?

**Comment by git-hulk** [comment]
@hy2014 I don't quite understand what you mean. Do you mean the data in the cloud are compacted and removed after moving them into the local instance?

**Reply by hy2014** [reply]
yes, after the local instance compact, my data is removed. so how can i keep my data?

**Reply by hy2014** [reply]
i found the log: Options.TTL= 30 days, so does kvrocks expire 30 days by default?

**Reply by git-hulk** [reply]
No, Kvrocks would never expire your data, as well as the compaction won't remove the data unless you set the expiration time. But I cannot identify the issue with your current information. 

Firstly, you can compare the sequence number via the command `INFO rocksdb` to see if it's close to your cloud instance.

**Reply by hy2014** [reply]
what is sequence num, i cannot find it. below is the "info rocksdb" in local:
`block_cache_usage:96
block_cache_pinned_usage[default]:96
estimate_keys[default]:0
index_and_filter_cache_usage[default]:0
level0_file_limit_slowdown[default]:0
level0_file_limit_stop[default]:0
pending_compaction_bytes_slowdown[default]:0
pending_compaction_bytes_stop[default]:0
level0_file_limit_stop_with_ongoing_compaction[default]:0
level0_file_limit_slowdown_with_ongoing_compaction[default]:0
memtable_count_limit_slowdown[default]:0
memtable_count_limit_stop[default]:0
estimate_keys[metadata]:0
index_and_filter_cache_usage[metadata]:0
level0_file_limit_slowdown[metadata]:0
level0_file_limit_stop[metadata]:0
pending_compaction_bytes_slowdown[metadata]:0
pending_compaction_bytes_stop[metadata]:0
level0_file_limit_stop_with_ongoing_compaction[metadata]:0
level0_file_limit_slowdown_with_ongoing_compaction[metadata]:0
memtable_count_limit_slowdown[metadata]:0
memtable_count_limit_stop[metadata]:0
estimate_keys[zset_score]:0
index_and_filter_cache_usage[zset_score]:0
level0_file_limit_slowdown[zset_score]:0
level0_file_limit_stop[zset_score]:0
pending_compaction_bytes_slowdown[zset_score]:0
pending_compaction_bytes_stop[zset_score]:0
level0_file_limit_stop_with_ongoing_compaction[zset_score]:0
level0_file_limit_slowdown_with_ongoing_compaction[zset_score]:0
memtable_count_limit_slowdown[zset_score]:0
memtable_count_limit_stop[zset_score]:0
estimate_keys[pubsub]:0
index_and_filter_cache_usage[pubsub]:0
level0_file_limit_slowdown[pubsub]:0
level0_file_limit_stop[pubsub]:0
pending_compaction_bytes_slowdown[pubsub]:0
pending_compaction_bytes_stop[pubsub]:0
level0_file_limit_stop_with_ongoing_compaction[pubsub]:0
level0_file_limit_slowdown_with_ongoing_compaction[pubsub]:0
memtable_count_limit_slowdown[pubsub]:0
memtable_count_limit_stop[pubsub]:0
estimate_keys[propagate]:0
index_and_filter_cache_usage[propagate]:0
level0_file_limit_slowdown[propagate]:0
level0_file_limit_stop[propagate]:0
pending_compaction_bytes_slowdown[propagate]:0
pending_compaction_bytes_stop[propagate]:0
level0_file_limit_stop_with_ongoing_compaction[propagate]:0
level0_file_limit_slowdown_with_ongoing_compaction[propagate]:0
memtable_count_limit_slowdown[propagate]:0
memtable_count_limit_stop[propagate]:0
estimate_keys[stream]:0
index_and_filter_cache_usage[stream]:0
level0_file_limit_slowdown[stream]:0
level0_file_limit_stop[stream]:0
pending_compaction_bytes_slowdown[stream]:0
pending_compaction_bytes_stop[stream]:0
level0_file_limit_stop_with_ongoing_compaction[stream]:0
level0_file_limit_slowdown_with_ongoing_compaction[stream]:0
memtable_count_limit_slowdown[stream]:0
memtable_count_limit_stop[stream]:0
estimate_keys[search]:0
index_and_filter_cache_usage[search]:0
level0_file_limit_slowdown[search]:0
level0_file_limit_stop[search]:0
pending_compaction_bytes_slowdown[search]:0
pending_compaction_bytes_stop[search]:0
level0_file_limit_stop_with_ongoing_compaction[search]:0
level0_file_limit_slowdown_with_ongoing_compaction[search]:0
memtable_count_limit_slowdown[search]:0
memtable_count_limit_stop[search]:0
block_cache_data_hit:0
block_cache_data_miss:0
block_cache_filter_hit:0
block_cache_filter_miss:0
block_cache_hit:0
block_cache_index_hit:0
block_cache_index_miss:0
block_cache_miss:0
all_mem_tables:14336
cur_mem_tables:14336
snapshots:0
num_immutable_tables:0
num_running_flushes:0
memtable_flush_pending:0
compaction_pending:0
num_running_compactions:0
num_live_versions:7
num_super_version:7
num_background_errors:0
flush_count:0
compaction_count:0
put_per_sec:0
get_per_sec:0
seek_per_sec:0
next_per_sec:0
prev_per_sec:0
is_bgsaving:no
is_compacting:no`

**Reply by hy2014** [reply]
i found local instance output log:

`,manifest_file_number is 1, next_file_number is 3, last_sequence is 0, log_number is 0,prev_log_number is 0,max_column_family is 0`

but my remote log:
`
Recovered from manifest file:/tmp/kvrocks/db/MANIFEST-000017 succeeded,manifest_file_number is 17, next_file_number is 192, last_sequence is 2707579, log_number is 189,prev_log_number is 0,max_column_family is 5,min_log_number_to_keep is 189
`

**Reply by git-hulk** [reply]
@hy2014 You can have a look at if your local instance db dir is correct. I belive it's a empty instance without any compaction and writes at all.

**Reply by hy2014** [reply]
1. first i copy the old db dir to local config dir path
`cp -r /home/xxx/kvrocks/db /var/lib/kvrocks/`
2. after copy and execute `ls -al`, the dir list bellow
```
-rw-r--r-- 1 root root 118179004 Aug 12 13:45 001849.sst
-rw-r--r-- 1 root root 136538912 Aug 12 13:45 001850.sst
-rw-r--r-- 1 root root 134467415 Aug 12 13:45 001851.sst
-rw-r--r-- 1 root root 136977938 Aug 12 13:45 001852.sst
-rw-r--r-- 1 root root 135687615 Aug 12 13:45 001853.sst
-rw-r--r-- 1 root root 135539202 Aug 12 13:45 001854.sst
-rw-r--r-- 1 root root    194006 Aug 12 13:45 001855.sst
-rw-r--r-- 1 root root  31185380 Aug 12 13:45 LOG
-rw-r--r-- 1 root root    255472 Aug 12 13:45 MANIFEST-00019
```
3. then remove the `LOG` file

4. then I start the kvrocks as follow command
`nohup kvrocks -c /var/lib/kvrocks/kvrocks.conf --dir /var/lib/kvrocks --pidfile /var/run/kvrocks/kvrocks.pid --bind 0.0.0.0 > /var/log/kvrocks/kvrocks.log 2>&1 &`

5. then the db dir like bellow
```
-rw-r--r-- 1 root root     0 Aug 12 13:50 001862.log
-rw-r--r-- 1 root root    16 Aug 12 13:50 CURRENT
-rw-r--r-- 1 root root    36 Aug 12 13:50 IDENTITY
-rw-r--r-- 1 root root     0 Aug 12 13:50 LOCK
-rw-r--r-- 1 root root 82168 Aug 12 13:50 LOG
-rw-r--r-- 1 root root 81554 Aug 12 13:50 LOG.old.1723470606254427
-rw-r--r-- 1 root root   633 Aug 12 13:50 MANIFEST-001863
-rw-r--r-- 1 root root 35454 Aug 12 13:50 OPTIONS-001861
-rw-r--r-- 1 root root 35982 Aug 12 13:50 OPTIONS-001865
drwxr-xr-x 2 root root  4096 Aug 12 13:50 archive

```
so. you can see, my data is lost, I check the LOG.old.17xxxxxxxxxx
```
2024/08/12-13:50:06.000687 79149 RocksDB version: 8.11.4
2024/08/12-13:50:06.000716 79149 Git sha 0
2024/08/12-13:50:06.000718 79149 Compile date 2024-07-06 08:01:13
2024/08/12-13:50:06.000723 79149 DB SUMMARY
2024/08/12-13:50:06.000725 79149 Host name (Env):  320c02f35009
2024/08/12-13:50:06.000727 79149 DB Session ID:  7Z64CUY1M55PTGOV1HF2
2024/08/12-13:50:06.000750 79149 MANIFEST file:  MANIFEST-000196 size: 255472 Bytes
2024/08/12-13:50:06.000752 79149 SST files in /var/lib/kvrocks/db dir, Total Num: 7, files: 001849.sst 001850.sst 001851.sst 00185
2.sst 001853.sst 001854.sst 001855.sst
2024/08/12-13:50:06.000754 79149 Write Ahead Log file in /var/lib/kvrocks/db:
.....  //Options arguments
2024/08/12-13:50:06.000932 79149 Fast CRC32 supported: Not supported on x86
2024/08/12-13:50:06.000934 79149 DMutex implementation: pthread_mutex_t
2024/08/12-13:50:06.017672 79149 [db/db_impl/db_impl_open.cc:327] Creating manifest 1
2024/08/12-13:50:06.047024 79149 [db/version_set.cc:6018] Recovering from manifest file: /var/lib/kvrocks/db/MANIFEST-000001
2024/08/12-13:50:06.047175 79149 [db/column_family.cc:616] --------------- Options for column family [default]:
...... // Options
2024/08/12-13:50:06.049430 79149 [db/version_set.cc:6061] Recovered from manifest file:/var/lib/kvrocks/db/MANIFEST-000001 succeeded,manifest_file_number is 1, next_file_number is 3, last_sequence is 0, log_number is 0,prev_log_number is 0,max_column_family is 0,min_log_number_to_keep is 0
2024/08/12-13:50:06.049442 79149 [db/version_set.cc:6076] Column family [default] (ID 0), log number is 0
2024/08/12-13:50:06.049497 79149 [db/db_impl/db_impl_open.cc:652] DB ID: cab4660c-d455-4423-8347-1151025a652e
2024/08/12-13:50:06.049630 79149 [db/version_set.cc:5511] Creating manifest 1857
2024/08/12-13:50:06.188678 79149 [db/db_impl/db_impl_open.cc:2180] SstFileManager instance 0x7feb11f87000
2024/08/12-13:50:06.188713 79149 DB pointer 0x7feb11fa3000
2024/08/12-13:50:06.189576 79169 [db/db_impl/db_impl.cc:988] ------- PERSISTING STATS -------
2024/08/12-13:50:06.189590 79169 [db/db_impl/db_impl.cc:1057] [Pre-GC] In-memory stats history size: 48 bytes, slice count: 0
2024/08/12-13:50:06.189592 79169 [db/db_impl/db_impl.cc:1066] [Post-GC] In-memory stats history size: 48 bytes, slice count: 0
2024/08/12-13:50:06.197837 79149 [db/column_family.cc:616] --------------- Options for column family [metadata]:
..... // Options
2024/08/12-13:50:06.233312 79149 [db/db_impl/db_impl.cc:3688] Created column family [search] (ID 6)
2024/08/12-13:50:06.253002 79149 [db/db_impl/db_impl.cc:487] Shutdown: canceling all background work
2024/08/12-13:50:06.253714 79149 [db/db_impl/db_impl.cc:681] Shutdown complete
```
also the LOG content is bellow
```
2024/08/12-13:50:06.254666 79149 RocksDB version: 8.11.4
2024/08/12-13:50:06.254706 79149 Git sha 0
2024/08/12-13:50:06.254708 79149 Compile date 2024-07-06 08:01:13
2024/08/12-13:50:06.254710 79149 DB SUMMARY
2024/08/12-13:50:06.254713 79149 Host name (Env):  320c02f35009
2024/08/12-13:50:06.254714 79149 DB Session ID:  7Z64CUY1M55PTGOV1HF3
2024/08/12-13:50:06.254774 79149 CURRENT file:  CURRENT
2024/08/12-13:50:06.254777 79149 IDENTITY file:  IDENTITY
2024/08/12-13:50:06.254780 79149 MANIFEST file:  MANIFEST-001857 size: 434 Bytes
2024/08/12-13:50:06.254783 79149 SST files in /var/lib/kvrocks/db dir, Total Num: 0, files: 
2024/08/12-13:50:06.254785 79149 Write Ahead Log file in /var/lib/kvrocks/db: 001856.log size: 0 ;
```
you can see sst file num become to zero. i do not know what happened?

**Reply by hy2014** [reply]
it seems kvrocks do not know my old MANIFEST file and create new MANIFEST?

**Reply by hy2014** [reply]
@git-hulk hope get some help,thx

**Reply by git-hulk** [reply]
@hy2014 Hmmmm... the root cause is that you're missing the `MANIFEST` file, so it will ignore old files and create from scratch.

**Reply by hy2014** [reply]
but old db exist MANIFEST file, MANIFEST-000196. why kvrocks cannot identify the MANIFEST file

**Reply by hy2014** [reply]
or i rename the MANIFEST-000196  to MANIFEST-000001, does it work?

**Reply by git-hulk** [reply]
No, it should be a file named `MANIFEST` and a pointer to `MANIFEST-00019`, I'm not sure why you lost that.

**Reply by hy2014** [reply]
but the new db dir also does not exist `MANIFEST` file.

**Reply by hy2014** [reply]
so you mean CURRENT?

**Reply by git-hulk** [reply]
Oh, yes. It's `CURRENT`.

**Reply by hy2014** [reply, marked as answer]
so i create and put my MANIFEST file name can resolve this problem?

**Reply by hy2014** [reply]
alread fix it,thx.
